docs/gguf.md

```diff
@@ -0,0 +1,511 @@
+    // The dimensions of the tensor.
+    uint32_t dimensions[n_dimensions];
+    // The number of elements in the tensor.
+    uint32_t n_elements;
+    // The type of the tensor.
+    ggml_type type;
+    // The offset of the tensor's data in this file in bytes.
+    // This offset is relative to `tensor_data`, not to the start
+    // of the file, to make it easier for writers to write the file.
+    // Readers should consider exposing this offset relative to the
+    // file to make it easier to read the data.
+    // Must be a multiple of `ALIGNMENT`.
+    uint64_t offset;
+};
+
+struct gguf_file_t {
+    // The header of the file.
+    gguf_header_t header;
+
+    // Tensor infos, which can be used to locate the tensor data.
+    gguf_tensor_info_t tensor_infos[header.tensor_count];
+
+    // Padding to the nearest multiple of `ALIGNMENT`.
+    uint8_t _padding[ALIGNMENT - (sizeof(header + tensor_infos) % ALIGNMENT)];
+
+    // Tensor data.
+    //
+    // This is arbitrary binary data corresponding to the weights of the model. This data should be close
+    // or identical to the data in the original model file, but may be different due to quantization or
+    // other optimizations for inference. Any such deviations should be recorded in the metadata or as
+    // part of the architecture definition.
+    //
+    // Each tensor's data must be stored within this array, and located through its `tensor_infos` entry.
+    // The offset of each tensor's data must be a multiple of `ALIGNMENT`, and the space between tensors
+    // should be padded to `ALIGNMENT` bytes.
+    uint8_t tensor_data[];
+};
+```
+
+## Standardized key-value pairs
+
+The following key-value pairs are standardized. This list may grow in the future as more use cases are discovered. Where possible, names are shared with the original model definitions to make it easier to map between the two.
+
+Not all of these are required, but they are all recommended. Keys that are required are bolded. For omitted pairs, the reader should assume that the value is unknown and either default or error as appropriate.
+
+The community can develop their own key-value pairs to carry additional data. However, these should be namespaced with the relevant community name to avoid collisions. For example, the `rustformers` community might use `rustformers.` as a prefix for all of their keys.
+
+If a particular community key is widely used, it may be promoted to a standardized key.
+
+### General
+
+#### Required
+
+- **`general.architecture: string`**: describes what architecture this model implements. All lowercase ASCII, with only `[a-z0-9]+` characters allowed. Known values include:
+  - `llama`
+  - `mpt`
+  - `gptneox`
+  - `gptj`
+  - `gpt2`
+  - `bloom`
+  - `falcon`
+  - `rwkv`
+- **`general.quantization_version: uint32`**: version of quantization scheme. Not required if the model is not quantized (i.e. no tensors are quantized). If any tensors are quantized, this _must_ be present.
```

Comment:
What is a quantization scheme and why is this needed in the model file?
The tensor quantization type is set by `gguf_tensor_info_t { ggml_type type; } `

Reply:
I think this was more informative. Look eg at the k-quants, they use a mix of different quantizations for the model.

Reply:
We already have a kv for file type aka quantization description `general.file_type`

Reply:
The defines for this in ggml.h is never used anywhere what I can tell?
https://github.com/ggerganov/ggml/blob/b57410a8b5b5c65306ed8d8986a2b7c1d43f3c1b/include/ggml/ggml.h#L202-L203

Reply:
It is used in ggml/examples to encode the quant version in the ftype param of the model file, but is never used for anything else than informative purposes in main.cpp .

I think this could be removed from the spec since that information should be in `general.file_type` like "mostly Q8_0" ?
Any executors could find out the full quantization scheme by looking at the tensor types.

Reply:
Found this PR https://github.com/ggerganov/llama.cpp/pull/1508. Why have a quantization version instead of simply adding a new quant type to the ggml types, in this case `Q4_0_scale_F16`. This way ggml could have support for both the previous version (renamed `Q4_0_scale_F32`) and the newer version?

Reply:
but yea, in the future.

Reply:
But if support should be removed just do as have been done to other quant types:
https://github.com/ggerganov/ggml/blob/b57410a8b5b5c65306ed8d8986a2b7c1d43f3c1b/include/ggml/ggml.h#L279-L280

Reply:
https://github.com/ggerganov/llama.cpp/pull/1405

see this as an example

Reply:
Yes, breaking changes to existing ggml types. What I meant was that we dont need a quantization version if we mark the changed types as not supported and make new types instead, increasing in the ggml type enum.

Reply:
The `GGML_QNT_VERSION` number (i.e. `general.quantization_version: uint32`) is used to mark breaking changes in the quantization formats / algorithms. The user code can check it to determine if a format is supported or not.

>  increasing in the ggml type enum.

I prefer to reuse the existing enum values. We don't worry about backwards compatibility before `ggml` v1.0 is released.

Reply:
The main problem with reusing enum and the quantization format names is that most users distribute model files with names like `llama-65b-q4_0.bin` so users dont know before downloading and testing them if it works or not. This is confusing and frustrating to end-users. This is also discussed in PR https://github.com/ggerganov/llama.cpp/pull/2434

Reply:
I included this as that's what the current version of GGML does, but (as mentioned in llama.cpp#2434) I would strongly prefer the breaking quantization changes to be treated as separate types, as klosax suggests. It's more descriptive, it means you only need one source of format version instead of two, and it's less error-prone.

If that's a possibility, I'd be a fan of removing the `quantization_version` value entirely.

Reply:
Maintaining backwards compatibility in user code is currently a rather complex task. In kobold.cpp they have resorted to [include](https://github.com/LostRuins/koboldcpp/tree/concedo/otherarch) several versions of ggml.c when trying to deal with the complexity.